edge/pkg/devicetwin/dmiserver/server.go

```diff
@@ -0,0 +1,263 @@
+	}
+
+	var deviceList []*pb.Device
+	var deviceModelList []*pb.DeviceModel
+	for _, device := range s.deviceList {
+		protocol, err := dtcommon.GetProtocolNameOfDevice(device)
+		if err != nil {
+			klog.Errorf("fail to get protocol name with err: %+v", err)
+			continue
+		}
+
+		if protocol == in.Mapper.Protocol {
+			dev, err := dtcommon.ConvertDevice(device)
+			if err != nil {
+				klog.Errorf("fail to convert device %s with err: %v", device.Name, err)
+				continue
+			}
+			dm, err := dtcommon.ConvertDeviceModel(s.deviceModelList[device.Spec.DeviceModelRef.Name])
+			if err != nil {
+				klog.Errorf("fail to convert device model %s with err: %v", s.deviceModelList[device.Spec.DeviceModelRef.Name], err)
+				continue
+			}
+			deviceList = append(deviceList, dev)
+			deviceModelList = append(deviceModelList, dm)
+		}
+	}
+
+	client, ok := dmiClientList[in.Mapper.Protocol]
+	if ok {
+		client.Conn.Close()
+		client.CancelFunc()
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(string(in.Mapper.Address))
+	if err != nil {
+		return nil, err
+	}
+	dmiClientList[in.Mapper.Protocol] = &DmiClient{
+		Client:     dc,
+		Ctx:        ctx,
+		Conn:       conn,
+		CancelFunc: cancelFunc,
+	}
+
+	return &pb.MapperRegisterResponse{
+		DeviceList: deviceList,
+		ModelList:  deviceModelList,
+	}, nil
+}
+
+func (s *server) ReportDeviceStatus(ctx context.Context, in *pb.ReportDeviceStatusRequest) (*pb.ReportDeviceStatusResponse, error) {
+	if !s.limiter.Allow() {
+		return nil, fmt.Errorf("fail to report device status because of too many request: %s", in.DeviceName)
+	}
+
+	for _, twin := range in.ReportedDevice.Twins {
+		propertyType, ok := twin.Reported.Metadata[PropertyType]
+		if !ok {
+			errLog := fmt.Sprintf("fail to get propertyType for property %s of device %s", twin.PropertyName, in.DeviceName)
+			klog.Errorf(errLog)
+			return nil, fmt.Errorf(errLog)
+		}
+		msg, err := CreateMessageTwinUpdate(twin.PropertyName, propertyType, twin.Reported.Value)
+		if err != nil {
+			klog.Errorf("fail to create message data for property %s of device %s with err: %v", twin.PropertyName, in.DeviceName, err)
+			return nil, err
+		}
+		handleDeviceTwin(in.DeviceName, msg)
+	}
+
+	return &pb.ReportDeviceStatusResponse{}, nil
+}
+
+func handleDeviceTwin(deviceName string, payload []byte) {
+	topic := dtcommon.DeviceETPrefix + deviceName + dtcommon.TwinETUpdateSuffix
+	target := modules.TwinGroup
+	resource := base64.URLEncoding.EncodeToString([]byte(topic))
+	// routing key will be $hw.<project_id>.events.user.bus.response.cluster.<cluster_id>.node.<node_id>.<base64_topic>
+	message := beehiveModel.NewMessage("").BuildRouter(modules.BusGroup, modules.UserGroup,
+		resource, messagepkg.OperationResponse).FillBody(string(payload))
+
+	beehiveContext.SendToGroup(target, *message)
+}
+
+// CreateMessageTwinUpdate create twin update message.
+func CreateMessageTwinUpdate(name string, valueType string, value string) (msg []byte, err error) {
```

Comment:
we do not suggest use named return value
```suggestion
func CreateMessageTwinUpdate(name, valueType, value string) ([]byte, error) {
```

Reply:
done